FWCore/Framework/interface/EventProcessor.h

```diff
@@ -213,42 +208,31 @@ namespace edm {
     void inputProcessBlocks();
     void endProcessBlock(bool cleaningUpAfterException, bool beginProcessBlockSucceeded);
 
-    void beginRun(ProcessHistoryID const& phid,
-                  RunNumber_t run,
-                  bool& globalBeginSucceeded,
-                  bool& eventSetupForInstanceSucceeded);
-    void endRun(ProcessHistoryID const& phid, RunNumber_t run, bool globalBeginSucceeded, bool cleaningUpAfterException);
-    void endUnfinishedRun(ProcessHistoryID const& phid,
-                          RunNumber_t run,
-                          bool globalBeginSucceeded,
-                          bool cleaningUpAfterException,
-                          bool eventSetupForInstanceSucceeded);
-
-    InputSource::ItemType processLumis(std::shared_ptr<void> const& iRunResource);
-    void endUnfinishedLumi();
-
-    void beginLumiAsync(edm::IOVSyncValue const& iSyncValue,
-                        std::shared_ptr<void> const& iRunResource,
-                        edm::WaitingTaskHolder iHolder);
-    void continueLumiAsync(edm::WaitingTaskHolder iHolder);
-
-    void handleEndLumiExceptions(std::exception_ptr const* iPtr, WaitingTaskHolder& holder);
-    void globalEndLumiAsync(edm::WaitingTaskHolder iTask, std::shared_ptr<LuminosityBlockProcessingStatus> iLumiStatus);
-    void streamEndLumiAsync(edm::WaitingTaskHolder iTask, unsigned int iStreamIndex);
+    InputSource::ItemType processRuns();
+    void beginRunAsync(IOVSyncValue const&, WaitingTaskHolder);
+    void releaseBeginRunResources(unsigned int iStreamIndex);
+    void endRunAsync(std::shared_ptr<RunProcessingStatus>, WaitingTaskHolder);
+    void handleEndRunExceptions(std::exception_ptr, WaitingTaskHolder const&);
+    void globalEndRunAsync(WaitingTaskHolder, std::shared_ptr<RunProcessingStatus>);
+    void streamEndRunAsync(WaitingTaskHolder, unsigned int iStreamIndex);
+    void endUnfinishedRun(bool cleaningUpAfterException);
+    void beginLumiAsync(IOVSyncValue const&, std::shared_ptr<RunProcessingStatus>, WaitingTaskHolder);
+    void continueLumiAsync(WaitingTaskHolder);
+    void handleEndLumiExceptions(std::exception_ptr, WaitingTaskHolder const&);
+    void globalEndLumiAsync(WaitingTaskHolder, std::shared_ptr<LuminosityBlockProcessingStatus>);
+    void streamEndLumiAsync(WaitingTaskHolder, unsigned int iStreamIndex);
+    void endUnfinishedLumi(bool cleaningUpAfterException);
     void readProcessBlock(ProcessBlockPrincipal&);
-    std::pair<ProcessHistoryID, RunNumber_t> readRun();
-    std::pair<ProcessHistoryID, RunNumber_t> readAndMergeRun();
+    void readRun(RunProcessingStatus&);
```

Comment:
Does this general rule apply for an object that exists before the function call and the function calls non-const functions on that object? I'm just trying to understand what the recommended style is in general.

In this particular case (readRun), I can make the change you suggest by returning a shared_ptr to a RunPrincipal and then having the calling code call the appropriate set function on the status object. I guess I will do this tomorrow. I'm not sure if I like it better...

It does make readRun different than readLuminosityBlock and readProcessBlock that already exist. And it would be hard to make the same change to readAndMergeRun.

Reply:
I made the change. The declaration is now:
```
std::shared_ptr<RunPrincipal> readRun();
```
And the one place where it is used is now:
```
status->setRunPrincipal(readRun());
```
instead of
```
readRun(*status);
```

Maybe more readable. If you want me to make other analogous changes for other functions, let me know. I'll push the commits for this first group of comments together when they are all done.